Testability Statements Section 5.3: BPI Service Orchestration (#213)
eea-oasis/baseline-standard: core/baseline-core-v1.0-psd01.md

* The expected outcome or result of the orchestrated actions is achieved.

#### **[R88]** 
BPI service orchestration utilized in a BPI MUST NOT introduce additional points of failure.

In the context of this document, a point of failure is defined as any non-redundant part of a BPI that, if not operational, would cause a BPI to fail.

[R88](#r88) Testability:

Preconditions:

* A BPI system is operational.
* The BPI service orchestration is properly configured and deployed.
* One or more sets of BPI test transactions. 

Test Steps:

1. Execute a set of BPI test transactions.
2. Monitor the behavior of the BPI service orchestration during the execution of each test transaction.
3. Verify that the BPI service orchestration behaves as expected for test transaction.
4. Simulate potential failure scenarios by deliberately disabling or interrupting specific components or services in the BPI.
5. Verify that the BPI service orchestration properly handles the failures and gracefully recovers without causing a complete BPI failure.
6. Simulate potential failure scenarios by deliberately disabling or interrupting the BPI service orchestration in the BPI.

> 7. Verify that BPI service orchestration failure is discvered and thet its built-in redundancies are successfully executed. 

Reviewer: ```suggestion
7. Verify that BPI service orchestration failure is discovered and that its built-in redundancies are successfully executed. 
```

[R93](#r93) Testability:

Preconditions:
* A BPI system is operational.
* The BPI service orchestration capability is properly configured and deployed.
* There are subscribers to the BPI service.

Test Steps:
1. A BPI service subscriber invokes a BPI service subscribed by multiple subscribers.
2. Monitor the execution of the BPI service operation.
3. Verify that the service request invokes a BPI service instance independent of the invoking subscriber.
4. Validate that only the invoking subscriber receives the expected responses or notifications from the BPI service invocation.
5. Modify the system state and trigger another invocation of the BPI service.
6. Verify that the service request invokes a BPI service instance independent of the invoking subscriber.
7. Validate that only the invoking subscriber receives the expected responses or notifications from the BPI service invocation.

Passing Criteria:
1. The BPI service orchestration ensures that the service request invokes BPI service instance independent of the invoking subscriber

> 2. Only the invoking subscriber receives the expected responses or notifications from the BPI service invocation..

Reviewer: ```suggestion
2. Only the invoking subscriber receives the expected responses or notifications from the BPI service invocation.
```

5. Ensure that that each invoked BPI service receives the necessary data from the data source through the BPI service orchestration independent from any of the other BPI service requests.

Passing Criteria:
1. The BPI service accesses the data source through the BPI service orchestration, and not directly.
2. Each invoked BPI service receives the necessary data from the data source through the BPI service orchestration independent from any of the other BPI service requests.

#### **[R96]** 
BPI service orchestration MUST not have a Material Impact on the overall system latency of the BPI.

[R96](#r96) Testability:

Preconditions:

* A BPI test system with a BPI service orchestration capability is set up and running.
* A test BPI service subscriber and BPI service consumer are set up within the BPI.
* The BPI has a defined threshold for "Material Impact".

Test Steps:

1. The BPI service subscriber invokes a BPI service.
2. Measure the time it takes for the BPI service request to propagate through the BPI service orchestration component and reach the BPI service consumer.
3. Compare the measured time with the defined "Material Impact" threshold.

Test Passing Criteria:

> * The test is considered passed if the measured time for the BPI service request to propagate through the BPI serviceorchestration component and reach the BPI service consumer is below the defined "Material Impact" threshold.

Reviewer: ```suggestion
* The test is considered passed if the measured time for the BPI service request to propagate through the BPI service orchestration component and reach the BPI service consumer is below the defined "Material Impact" threshold.
```
